x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/oidc/TransportOpenIdConnectLogoutAction.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.authc.Realms;
+import org.elasticsearch.xpack.security.authc.TokenService;
+import org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectRealm;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Map;
+
+/**
+ * Transport action responsible for generating an OpenID connect logout request to be sent to an OpenID Connect Provider
+ */
+public class TransportOpenIdConnectLogoutAction extends HandledTransportAction<OpenIdConnectLogoutRequest, OpenIdConnectLogoutResponse> {
+
+    private final Realms realms;
+    private final TokenService tokenService;
+    private static final Logger logger = LogManager.getLogger(TransportOpenIdConnectLogoutAction.class);
+
+    @Inject
+    public TransportOpenIdConnectLogoutAction(TransportService transportService, ActionFilters actionFilters, Realms realms,
+                                              TokenService tokenService) {
+        super(OpenIdConnectLogoutAction.NAME, transportService, actionFilters,
+            (Writeable.Reader<OpenIdConnectLogoutRequest>) OpenIdConnectLogoutRequest::new);
+        this.realms = realms;
+        this.tokenService = tokenService;
+    }
+
+    @Override
+    protected void doExecute(Task task, OpenIdConnectLogoutRequest request, ActionListener<OpenIdConnectLogoutResponse> listener) {
+        invalidateRefreshToken(request.getRefreshToken(), ActionListener.wrap(ignore -> {
+            try {
+                final String token = request.getToken();
+                tokenService.getAuthenticationAndMetaData(token, ActionListener.wrap(
+                    tuple -> {
+                        final Authentication authentication = tuple.v1();
+                        final Map<String, Object> tokenMetadata = tuple.v2();
+                        validateAuthenticationAndMetadata(authentication, tokenMetadata);
+                        tokenService.invalidateAccessToken(token, ActionListener.wrap(
+                            result -> {
+                                if (logger.isTraceEnabled()) {
+                                    logger.trace("OpenID Connect Logout for user [{}] and token [{}...{}]",
+                                        authentication.getUser().principal(),
+                                        token.substring(0, 8),
+                                        token.substring(token.length() - 8));
+                                }
+                                OpenIdConnectLogoutResponse response = buildResponse(authentication, tokenMetadata);
+                                listener.onResponse(response);
+                            }, listener::onFailure)
+                        );
+                    }, listener::onFailure));
+            } catch (IOException e) {
+                logger.debug("Internal error during OpenID Connect Logout");
```

Comment:
tracing this back to `TokenService#decodeToke` (which throws **and** has a listener... _Grrr_) it looks like `IOException` is actually a sign of an invalid token, so I would drop this log line because the error is not internal.